### PR TITLE
Build a simple event loop for collective.

### PR DESCRIPTION
**Pull request by trivialfis**

### CODE REVIEW

**Comment by trivialfis** [comment]
cc @rongou .

File: tests/cpp/collective/test_loop.cc [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#include <chrono>        // for seconds
+#include <cinttypes>     // for int8_t
+#include <memory>        // for make_shared
+#include <system_error>  // make_error_code, errc
+
+#include "../../../src/collective/loop.h"
+
+namespace xgboost::collective {
+TEST(Loop, Timeout) {
+  system::SocketStartup();
```
Comment (rongou) [inline tests/cpp/collective/test_loop.cc:15]: Maybe put this in `SetUp()`.

File: tests/cpp/collective/test_loop.cc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  auto loop = std::make_shared<Loop>(timeout);
+
+  TCPSocket sock;
+  std::vector<std::int8_t> data(1);
+  Loop::Op op{Loop::Op::kRead, 0, data.data(), data.size(), &sock, 0};
+  loop->Submit(op);
+  auto rc = loop->Block();
+  ASSERT_FALSE(rc.OK());
+  ASSERT_EQ(rc.Code(), std::make_error_code(std::errc::timed_out));
+  system::SocketFinalize();
```
Comment (rongou) [inline tests/cpp/collective/test_loop.cc:26]: And this in `TearDown()`.

**Review by rongou — APPROVED** [review]

**Comment by trivialfis** [comment]
@rongou Made some additional changes to fix the tests on macos and windows.

**Review by rongou — APPROVED** [review]

**Pull request merged**